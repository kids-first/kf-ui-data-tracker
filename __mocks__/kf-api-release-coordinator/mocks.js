import {GET_STUDY_RELEASES, GET_RELEASED_STUDY} from '../../src/state/queries';
import getStudyReleases from './responses/getStudyReleases.json';
import getStudyReleasesEmpty from './responses/getStudyReleasesEmpty.json';
import allReleaseStudies from './responses/allReleaseStudies.json';

export const coordMocks = {
  getStudyReleases: {
    request: {
      query: GET_STUDY_RELEASES,
      variables: {
        id: 'U3R1ZHlOb2RlOlNEXzhXWDhRUTA2',
      },
    },
    result: getStudyReleases,
  },
  getStudyReleasesError: {
    request: {
      query: GET_STUDY_RELEASES,
      variables: {
        id: 'U3R1ZHlOb2RlOlNEXzhXWDhRUTA2',
      },
    },
    error: new Error('Failed to fetch releases information'),
  },
  getStudyReleasesEmpty: {
    request: {
      query: GET_STUDY_RELEASES,
      variables: {
        id: 'U3R1ZHlOb2RlOlNEXzhXWDhRUTA2',
      },
    },
    result: getStudyReleasesEmpty,
  },
  allReleaseStudies: {
    request: {
      query: GET_RELEASED_STUDY,
    },
    result: allReleaseStudies,
  },
  allReleaseStudiesError: {
    request: {
      query: GET_RELEASED_STUDY,
    },
    error: new Error('Failed to fetch study releases information'),
  },
};
