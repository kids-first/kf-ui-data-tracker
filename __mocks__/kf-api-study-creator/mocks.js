import allStudies from './responses/allStudies';
import studyByKfId from './responses/studyByKfId.json';
import deleteFile from './responses/deleteFile.json';
import fileByKfId from './responses/fileByKfId.json';
import myProfile from './responses/myProfile.json';
import {
  ALL_STUDIES,
  GET_STUDY_BY_ID,
  MY_PROFILE,
} from '../../src/state/queries';
import {GET_FILE_BY_ID} from '../../src/documents/queries';
import updateFile from './responses/updateFile.json';
import updateVersion from './responses/updateVersion.json';
import createVersion from './responses/createVersion.json';
import {
  DELETE_FILE,
  UPDATE_FILE,
  UPDATE_VERSION,
  CREATE_VERSION,
} from '../../src/documents/mutations';

export const mocks = [
  {
    request: {
      query: ALL_STUDIES,
    },
    result: allStudies,
  },
  {
    request: {
      query: GET_STUDY_BY_ID,
      variables: {
        kfId: 'SD_8WX8QQ06',
      },
    },
    result: studyByKfId,
  },
  {
    request: {
      query: GET_STUDY_BY_ID,
      variables: {
        kfId: 'SD_8WX8QQ06',
      },
    },
    error: new Error('something went wrong'),
  },
  {
    request: {
      query: DELETE_FILE,
      variables: {
        kfId: 'SF_Y07IN1HO',
      },
    },
    result: deleteFile,
  },
  {
    request: {
      query: GET_FILE_BY_ID,
      variables: {
        kfId: 'SF_5ZPEM167',
      },
    },
    result: fileByKfId,
  },
  {
    request: {
      query: UPDATE_FILE,
      variables: {
        kfId: 'SF_5ZPEM167',
        name: 'foo bar file',
        description: 'Some description here',
        fileType: 'CLN',
      },
    },
    result: updateFile,
  },
  {
    request: {
      query: UPDATE_VERSION,
      variables: {
        versionId: 'FV_DMTKEFQW',
        description: 'version description',
        state: 'APP',
      },
    },
    result: updateVersion,
  },

  {
    request: {
      query: CREATE_VERSION,
      variables: {
        file: {
          lastModified: 1567797130510,
          name: 'organize.tsv',
          size: 7,
          type: 'text/tab-separated-values',
          webkitRelativePath: '',
        },
        fileId: 'SF_5ZPEM167',
        description: 'version description',
      },
    },
    result: createVersion,
  },
  {
    request: {
      query: MY_PROFILE,
    },
    result: myProfile,
  },
];
