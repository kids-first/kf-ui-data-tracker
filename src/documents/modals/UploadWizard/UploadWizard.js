import React, {useState, useEffect} from 'react';
import PropTypes from 'prop-types';
import {useMutation} from '@apollo/react-hooks';
import {Button, Header, Modal, Icon, Message} from 'semantic-ui-react';
import {
  ChooseMethodStep,
  DocumentSelectionStep,
  SuccessStep,
  VersionSummaryStep,
} from './UploadSteps';
import {GET_STUDY_BY_ID} from '../../../state/queries';
import {CREATE_VERSION} from '../../mutations';
import {sortFilesBySimilarity} from '../../utilities';

// store all of our upload steps and their associated components
const UPLOAD_STEPS = {
  0: {
    title: 'Choose an Upload Method',
    comp: ChooseMethodStep,
  },
  1: {
    title: 'Update Existing Study Document',
    comp: DocumentSelectionStep,
  },
  2: {
    title: 'Summarize Your Update',
    comp: VersionSummaryStep,
  },
  3: {
    title: 'Document Updated',
    comp: SuccessStep,
  },
};

/** custom hook to set a 3 second timer  */
const useTimerHook = (isActive, onEnd) => {
  const [seconds, setSeconds] = useState(0);

  useEffect(() => {
    let interval = null;
    if (isActive && seconds < 3) {
      interval = setInterval(() => {
        setSeconds(seconds => seconds + 1);
      }, 1000);
    } else if (seconds >= 3) {
      clearInterval(interval);
      onEnd();
    } else {
      clearInterval(interval);
    }
    return () => clearInterval(interval);
  }, [isActive, seconds, onEnd]);

  return seconds;
};

/** Multi-step modal for uplaoding files as document versions or creating new documents */
const UploadWizard = ({
  onCloseDialog,
  history,
  file,
  startingStep = 0,
  fileList,
  studyId,
  allowUploadFile,
  allowUploadVersion,
}) => {
  const [createVersion] = useMutation(CREATE_VERSION, {
    refetchQueries: [{query: GET_STUDY_BY_ID, variables: {kfId: studyId}}],
  });

  // The current step that the flow is on
  const [step, setStep] = useState(startingStep);
  // store the selected study file to create version for
  const [fileToUpdate, setFileToUpdate] = useState(null);

  // TODO: set upload error state in ui
  // For any errors that occur during upload
  // const [errors, setErrors] = useState();

  // For the uploading stage
  const [onUploading, setUploading] = useState(false);
  // Hold the version change description
  const [description, setDescription] = useState();

  const similarDocuments = sortFilesBySimilarity(file, fileList);

  const handleCloseDialog = () => {
    onCloseDialog();
  };

  const [isTimerActive, setIsTimerActive] = useState(false);
  const seconds = useTimerHook(isTimerActive, () => {
    handleCloseDialog();
    if (history.location.pathname.split('/').includes('new-document'))
      history.push(`/study/${studyId}/documents/${fileToUpdate.kfId}`);
  });

  /** TODO: abstract to custom hook or external util */
  const handleSave = async props => {
    try {
      setUploading(true);
      await createVersion({
        variables: {file, fileId: fileToUpdate.kfId, description},
      });
      // TODO: maybe make this part of local state
      sessionStorage.setItem('kf_updated_docs', fileToUpdate.kfId);

      setUploading(false);
      setIsTimerActive(true);
      setStep(3);
    } catch (e) {
      console.error(e);
      setIsTimerActive(false);
      // setErrors(e)
    }
  };

  return (
    <Modal
      open={true}
      onClose={() => {
        handleCloseDialog();
      }}
      closeIcon
    >
      <Header>
        <Header.Subheader as="h2" data-testid="wizard-subhead">
          <Icon name="upload cloud" />
          Upload: {file.name}
        </Header.Subheader>
        {UPLOAD_STEPS[step].title}
      </Header>

      <Modal.Content>
        {similarDocuments.matches.length && step < 2 ? (
          <Message info size="mini">
            <Icon name="info circle" />
            <strong>{similarDocuments.matches.length}</strong> similar documents
            found
          </Message>
        ) : null}

        {UPLOAD_STEPS[step].comp({
          history,
          file,
          fileList,
          setStep,
          setFileToUpdate,
          fileToUpdate,
          setDescription,
          handleCloseDialog,
          isTimerActive,
          seconds,
          allowUploadFile,
          allowUploadVersion,
        })}
      </Modal.Content>

      <Modal.Actions>
        <Button
          icon
          labelPosition="left"
          floated="left"
          size="mini"
          disabled={onUploading}
          onClick={() => {
            if (step === 0 || step === 3) {
              handleCloseDialog();
            }
            setStep(step - 1);
          }}
        >
          {step === 3 ? (
            <>
              <Icon name="x" />
              Close
            </>
          ) : (
            <>
              <Icon name="arrow left" /> Back{' '}
            </>
          )}
        </Button>

        <Button
          primary
          icon
          labelPosition="left"
          size="mini"
          disabled={step !== 2 || !file}
          data-testid="upload-button"
          onClick={handleSave}
        >
          <Icon name="upload cloud" />
          {onUploading ? 'UPLOADING ...' : 'UPLOAD'}
        </Button>
      </Modal.Actions>
    </Modal>
  );
};

UploadWizard.propTypes = {
  /** Function to be called when the modal should be closed */
  onCloseDialog: PropTypes.func.isRequired,
  /** The file that the document is being created or updated for */
  file: PropTypes.object.isRequired,
  /** Array of file nodes for current study  */
  fileList: PropTypes.array.isRequired,
  /** React-router hisotry object  */
  history: PropTypes.any,
  /** Mutation that will upload and create the version */
  createVersion: PropTypes.func,
};

export default UploadWizard;
