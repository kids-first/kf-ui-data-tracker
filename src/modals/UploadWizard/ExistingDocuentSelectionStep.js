import React from 'react';
import { Icon, Table } from 'semantic-ui-react';
import * as stringSimilarity from 'string-similarity';

const DocRow = ({ document, setStep, setFileToUpdate }) => {
  const isSimilar = document.rating > 0.3
  return (<Table.Row style={{ background: isSimilar ? '#f8ffff' : 'inherit' }} onClick={() => { setFileToUpdate(document); setStep(2); }} >
    <Table.Cell>
      <Icon name='file' />  {document.name}
    </Table.Cell>
  </Table.Row>
  )
}

const SimilarDocumentsTable = ({ documents, setStep, setFileToUpdate }) => (
  <Table stackable selectable compact color="blue" >
    <Table.Header>
      <Table.Row>
        <Table.HeaderCell>Select a Study Document to Update</Table.HeaderCell>
      </Table.Row>
    </Table.Header>
    <Table.Body>
      {
        documents.reverse().map(doc => (
          <DocRow document={doc} setStep={setStep} setFileToUpdate={setFileToUpdate} />
        ))
      }
    </Table.Body>
  </Table >
)


const ExistingDocuentSelectionStep = ({ fileList, file, setStep, setFileToUpdate }) => {

  // find bestMatches for filename's similar to the uploaded document
  const fileMatches = stringSimilarity.findBestMatch(file.name || '', fileList.length ? fileList.map(x => x.node.name) : [])
  // sort all study files by bestMatches rating
  const updateDocumentsList = fileList.map(({ node }) => ({ rating: fileMatches.ratings.filter(({ target }) => target === node.name)[0].rating, ...node })).sort((a, b) => (a.rating > b.rating) ? 1 : -1)

  return (<>
    <p>Adding changes to existing documents in your study? You can rest easy knowing that any previous versions of your documents are automatically archived, and available to you for easy download/review at any time.</p>
    <SimilarDocumentsTable documents={updateDocumentsList} setStep={setStep} setFileToUpdate={setFileToUpdate} />
  </>
  )
};

export default ExistingDocuentSelectionStep;