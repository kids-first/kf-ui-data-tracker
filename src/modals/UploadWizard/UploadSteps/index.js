export {default as ChooseMethodStep} from './ChooseMethodStep';
export {default as DocumentSelectionStep} from './DocumentSelectionStep';
export {default as VersionSummaryStep} from './VersionSummaryStep';
export {default as SuccessStep} from './SuccessStep';
