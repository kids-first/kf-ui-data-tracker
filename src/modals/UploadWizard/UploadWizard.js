import React, {useState, useEffect} from 'react';
import PropTypes from 'prop-types';
import {Button, Header, Modal, Icon, Message} from 'semantic-ui-react';
import {
  ChooseMethodStep,
  DocumentSelectionStep,
  SuccessStep,
  VersionSummaryStep,
} from './UploadSteps';
import {graphql} from 'react-apollo';

import {GET_STUDY_BY_ID} from '../../state/queries';
import {CREATE_VERSION} from '../../state/mutations';
import {sortFilesBySimilarity} from '../../common/fileUtils';

// store all of our upload steps and their associated components
const UPLOAD_STEPS = {
  0: {
    title: 'Choose an Upload Method',
    comp: ChooseMethodStep,
  },
  1: {
    title: 'Update Existing Study Document',
    comp: DocumentSelectionStep,
  },
  2: {
    title: 'Summarize Your Update',
    comp: VersionSummaryStep,
  },
  3: {
    title: 'Document Updated',
    comp: SuccessStep,
  },
};

/** custom hook to set a 3 second timer  */
const useTimerHook = (isActive, onEnd) => {
  const [seconds, setSeconds] = useState(0);

  useEffect(() => {
    let interval = null;
    if (isActive && seconds < 3) {
      interval = setInterval(() => {
        setSeconds(seconds => seconds + 1);
      }, 1000);
    } else if (seconds >= 3) {
      clearInterval(interval);
      onEnd();
    } else {
      clearInterval(interval);
    }
    return () => clearInterval(interval);
  }, [isActive, seconds, onEnd]);

  return seconds;
};

/** Multi-step modal for uplaoding files as document versions or creating new documents */
const UploadWizard = ({
  createVersion,
  onCloseDialog,
  history,
  file,
  fileList,
  studyId,
}) => {
  // The current step that the flow is on
  const [step, setStep] = useState(0);
  // store the selected study file to create version for
  const [fileToUpdate, setFileToUpdate] = useState(null);

  // TODO: set upload error state in ui
  // For any errors that occur during upload
  // const [errors, setErrors] = useState();

  // For the uploading stage
  const [onUploading, setUploading] = useState(false);
  // Hold the version change description
  const [description, setDescription] = useState();

  const similarDocuments = sortFilesBySimilarity(file, fileList);

  const handleCloseDialog = () => {
    onCloseDialog();
  };

  const [isTimerActive, setIsTimerActive] = useState(false);
  const seconds = useTimerHook(isTimerActive, handleCloseDialog);

  /** TODO: abstract to custom hook or external util */
  const handleSave = async props => {
    try {
      setUploading(true);
      await createVersion({
        variables: {file, fileId: fileToUpdate.kfId, description},
      });
      // TODO: maybe make this part of local state
      sessionStorage.setItem('kf_updated_docs', fileToUpdate.kfId);

      setUploading(false);
      setIsTimerActive(true);
      setStep(3);
    } catch (e) {
      console.error(e);
      setIsTimerActive(false);
      // setErrors(e)
    }
  };

  return (
    <Modal
      open={true}
      onClose={() => {
        handleCloseDialog();
      }}
      closeIcon
    >
      <Header>
        <Header.Subheader as="h2" data-testid="wizard-subhead">
          <Icon name="upload cloud" />
          Upload: {file.name}
        </Header.Subheader>
        {UPLOAD_STEPS[step].title}
      </Header>

      <Modal.Content>
        {similarDocuments.matches.length && step < 2 ? (
          <Message info size="mini">
            <Icon name="info circle" />
            <strong>{similarDocuments.matches.length}</strong> similar documents
            found
          </Message>
        ) : null}

        {UPLOAD_STEPS[step].comp({
          history,
          file,
          fileList,
          setStep,
          setFileToUpdate,
          fileToUpdate,
          setDescription,
          handleCloseDialog,
          isTimerActive,
          seconds,
        })}
      </Modal.Content>

      <Modal.Actions>
        <Button
          icon
          labelPosition="left"
          floated="left"
          size="mini"
          disabled={onUploading}
          onClick={() => {
            if (step === 0 || step === 3) {
              handleCloseDialog();
            }
            setStep(step - 1);
          }}
        >
          {step === 3 ? (
            <>
              <Icon name="x" />
              Close
            </>
          ) : (
            <>
              <Icon name="arrow left" /> Back{' '}
            </>
          )}
        </Button>

        <Button
          primary
          icon
          labelPosition="left"
          size="mini"
          disabled={step !== 2 || !file}
          data-testid="upload-button"
          onClick={handleSave}
        >
          <Icon name="upload cloud" />
          {onUploading ? 'UPLOADING ...' : 'UPLOAD'}
        </Button>
      </Modal.Actions>
    </Modal>
  );
};

UploadWizard.propTypes = {
  /** Function to be called when the modal should be closed */
  onCloseDialog: PropTypes.func.isRequired,
  /** The file that the document is being created or updated for */
  file: PropTypes.object.isRequired,
  /** Array of file nodes for current study  */
  fileList: PropTypes.array.isRequired,
  /** React-router hisotry object  */
  history: PropTypes.any,
  /** Mutation that will upload and create the version */
  createVersion: PropTypes.func.isRequired,
};

export default graphql(CREATE_VERSION, {
  name: 'createVersion',
  options: ({studyId}) => ({
    refetchQueries: [{query: GET_STUDY_BY_ID, variables: {kfId: studyId}}],
  }),
})(UploadWizard);

// export default UploadWizard;
