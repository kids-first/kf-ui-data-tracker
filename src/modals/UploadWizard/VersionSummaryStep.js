import React from 'react';
import PropTypes from 'prop-types';
import { Header, Icon, Divider, Grid, Segment, Form } from 'semantic-ui-react';


/**
 * Step 2 in multi-step upload modal. 
 * Describes a creates a document version
 */
const VersionSummaryStep = ({ fileToUpdate, file }) => (
  <>
    <Segment size="mini" basic>
      <Grid columns={2} stackable textAlign='center'>
        <Divider vertical>with</Divider>
        <Grid.Row >
          <Grid.Column>
            <Header icon> <Icon name='copy' />  Update Document:
            <Header.Subheader>{fileToUpdate.name}</Header.Subheader>
            </Header>

          </Grid.Column>

          <Grid.Column>
            <Header icon> <Icon name='upload cloud' />  Uploaded Document:
              <Header.Subheader>{file.name}</Header.Subheader>
            </Header>
          </Grid.Column>
        </Grid.Row>
      </Grid>
    </Segment>


    <Segment color="blue" raised >
      <h5>
        <Icon name='pencil' />
        Help keep track of your document history by telling us what may have
        changed in this version
      </h5>
      <Form>
        <Form.Field required focus>
          <label>Summarize Document Changes:</label>
          <Form.TextArea
            focus
            data-testid="description-input"
            name="description"
            type="text"
          />
        </Form.Field>
      </Form>
    </Segment>

  </>
)

VersionSummaryStep.propTypes = {
  /** The file that the document is being created or updated for */
  file: PropTypes.object.isRequired,
  /** Existing file node from study */
  fileToUpdate: PropTypes.object.isRequired,
}

export default VersionSummaryStep;