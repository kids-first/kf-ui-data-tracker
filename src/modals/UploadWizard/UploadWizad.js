import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Button, Header, Modal, Icon, Message } from 'semantic-ui-react';
import ChooseMethodStep from './ChooseMethodStep';
import ExistingDocuentSelectionStep from './ExistingDocuentSelectionStep'
import VersionSummaryStep from './VersionSummaryStep';
import * as stringSimilarity from 'string-similarity';


const UPLOAD_STEPS = {
  0: {
    title: 'Choose an Upload Method',
    comp: ChooseMethodStep,
  },
  1: {
    title: 'Update Existing Study Document',
    comp: ExistingDocuentSelectionStep
  },
  2: {
    title: 'Summarize Your Update',
    comp: VersionSummaryStep,
  }
}

/** Multi-step modal for uplaoding new documents as files versions or creating new documents */
const UploadWizard = ({ onCloseDialog, history, file, fileList }) => {
  // The current step that the flow is on
  const [step, setStep] = useState(0);
  const [fileToUpdate, setFileToUpdate] = useState(null);


  const MIN_SIMILIRATIY = 0.3;
  // find bestMatches for filename's similar to the uploaded document
  const fileMatches = stringSimilarity.findBestMatch(file.name || '', fileList.length ? fileList.map(x => x.node.name) : [])
  // sort bestMatches by rating descending
  const similarDocuments = fileMatches.ratings.filter(f => f.rating > MIN_SIMILIRATIY).sort((a, b) => (a.rating > b.rating) ? 1 : -1).reverse();


  return (
    <Modal open={true} onClose={() => { setStep(0); onCloseDialog() }} closeIcon>

      <Header>
        <Header.Subheader as='h2'><Icon name="upload cloud" />Upload: {file.name}</Header.Subheader>
        {UPLOAD_STEPS[step].title}
      </Header>

      <Modal.Content >

        {similarDocuments.length && step < 2 ?
          <Message info size="mini" >
            <Icon name='info circle' />
            <strong>{similarDocuments.length}</strong> similar documents found
          </Message>
          : null}

        {UPLOAD_STEPS[step].comp({ history, file, fileList, setStep, setFileToUpdate, fileToUpdate })}


      </Modal.Content>

      <Modal.Actions>

        <Button
          icon
          labelPosition="left"
          floated="left"
          size="mini"
          onClick={() => {
            if (step === 0) {
              setStep(0);
              onCloseDialog();
            }
            setStep(step - 1);
          }}
        >
          <Icon name="arrow left" />
          Back
          </Button>

        <Button
          primary
          icon
          labelPosition="left"
          size="mini"
          disabled={step < 2 || !file}
        // onClick={handleSave}
        >
          <Icon name="upload cloud" />
          UPLOAD
        </Button>
      </Modal.Actions>

    </Modal >
  );

};

UploadWizard.propTypes = {
  /** Function to be called when the modal should be closed */
  onCloseDialog: PropTypes.func.isRequired,
  /** The file that the document is being created or updated for */
  file: PropTypes.object.isRequired,
  /** Array of file nodes for current study  */
  fileList: PropTypes.array.isRequired,
  /** React-router hisotry object  */
  history: PropTypes.any,
}

export default UploadWizard;