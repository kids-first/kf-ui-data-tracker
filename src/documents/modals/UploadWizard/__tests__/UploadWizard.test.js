import React from 'react';
import wait from 'waait';
import {MockedProvider} from 'react-apollo/test-utils';
import {MemoryRouter} from 'react-router-dom';
import {render, fireEvent, cleanup, act} from 'react-testing-library';
import UploadWizard from '../UploadWizard';
import {mocks} from '../../../../../__mocks__/kf-api-study-creator/mocks';
import studyByKfId from '../../../../../__mocks__/kf-api-study-creator/responses/studyByKfId.json';

afterEach(cleanup);

const handleClose = jest.fn();
const studyFiles = studyByKfId.data.studyByKfId.files.edges;
const uploadedFile = new File([''], 'organize.tsv', {
  type: 'text/tab-separated-values',
  lastModified: new Date(0),
});
const historyMock = {
  push: jest.fn(),
  location: {pathname: '/study/SD_8WX8QQ06/documents'},
};
let file = {
  lastModified: 1567797130510,
  name: 'organize.tsv',
  size: 7,
  type: 'text/tab-separated-values',
  webkitRelativePath: '',
};
const sleep = ms => new Promise(resolve => setTimeout(() => resolve(), ms));

it('renders with uploaded file name', async () => {
  const {container, getByText} = render(
    <MockedProvider mocks={mocks}>
      <UploadWizard
        history={historyMock}
        studyId={studyByKfId.data.studyByKfId.kfId}
        file={uploadedFile}
        fileList={studyFiles}
        onCloseDialog={handleClose}
      />
    </MockedProvider>,
  );

  await wait(0);

  //react portal w/ modal is added to document
  expect(document.querySelectorAll('.ui.modal').length).toBe(1);
  // renders the uploaded file name in the header
  expect(getByText(/organize\.tsv/i)).toBeTruthy();

  expect(container).toMatchSnapshot();
});

it('progresses to Step 1: Update Existing Study Document ', async () => {
  const {getByText, getByTestId, queryAllByTestId} = render(
    <MockedProvider mocks={mocks}>
      <UploadWizard
        history={historyMock}
        studyId={studyByKfId.data.studyByKfId.kfId}
        file={uploadedFile}
        fileList={studyFiles}
        onCloseDialog={handleClose}
      />
    </MockedProvider>,
  );

  await wait(0);

  act(() => {
    fireEvent.click(getByTestId('update-existing-button'));
  });

  await wait(0);

  expect(getByText(/Update Existing Study Document/i)).toBeTruthy();
  // should render a list of two documents
  expect(queryAllByTestId(/document-item/).length).toBe(studyFiles.length);
  // should only have one similar document in list
  expect(getByTestId('similar-document-item')).toBeDefined();
});

it('progresses to Step 2: Summarize Your Update', async () => {
  const {getByText, getByTestId} = render(
    <MockedProvider mocks={mocks}>
      <UploadWizard
        history={historyMock}
        studyId={studyByKfId.data.studyByKfId.kfId}
        file={uploadedFile}
        fileList={studyFiles}
        onCloseDialog={handleClose}
      />
    </MockedProvider>,
  );

  await wait(0);

  // choose update existing document
  act(() => {
    fireEvent.click(getByTestId('update-existing-button'));
  });

  await wait(0);

  // select the first document in the list "organize.jpeg"
  act(() => {
    fireEvent.click(getByTestId('similar-document-item'));
  });

  await wait(0);
  /* make sure we're on the right step */
  expect(getByText(/Summarize Your Update/i)).toBeTruthy();
  /* make sure we're updating the correct document */
  expect(getByText(/organization\.jpeg/i)).toBeTruthy();
});

it('progresses to Step 3: Document Updated ', async done => {
  const {getByText, getByTestId} = render(
    <MockedProvider mocks={mocks}>
      <MemoryRouter initialEntries={['/study/SD_8WX8QQ06/documents']}>
        <UploadWizard
          history={historyMock}
          studyId={studyByKfId.data.studyByKfId.kfId}
          file={file}
          fileList={studyFiles}
          onCloseDialog={handleClose}
        />
      </MemoryRouter>
    </MockedProvider>,
  );

  await wait(0);

  act(() => {
    fireEvent.click(getByTestId('update-existing-button'));
  });

  await wait(0);

  // select the first document in the list
  act(() => {
    fireEvent.click(getByTestId('similar-document-item'));
  });

  await wait(0);
  const inputVal = 'version description';
  act(() => {
    fireEvent.change(getByTestId('description-input'), {
      target: {value: inputVal},
    });
  });

  // make sure our description is input correctly
  expect(getByTestId('description-input').value).toBe(inputVal);

  act(() => {
    fireEvent.click(getByTestId('upload-button'));
  });
  await wait(10);

  expect(handleClose.mock.calls.length).toBe(0);
  await wait(10);
  expect(getByText(/Document Updated/i)).toBeTruthy();

  // make sure we progress the to sucess screen
  expect(getByText('Updated Document organization.jpeg')).toBeTruthy();
  // make sure close timer begins
  expect(getByText(/Exit \(3\)/i)).toBeTruthy();

  // wait 1 sec and make sure timer is counting down
  await sleep(1000);
  expect(getByText(/Exit \(2\)/i)).toBeTruthy();
  expect(handleClose.mock.calls.length).toBe(0);
  await sleep(1000);
  expect(getByText(/Exit \(1\)/i)).toBeTruthy();
  await sleep(1000);
  // make sure it closes after 3 seconds
  expect(handleClose.mock.calls.length).toBe(1);

  done();
}, 6000);
