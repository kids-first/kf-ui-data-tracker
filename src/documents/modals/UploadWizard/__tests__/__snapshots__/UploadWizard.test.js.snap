// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with uploaded file name 1`] = `<div />`;
